### PR TITLE
[WIP] Try manipulating the DOM to add the 'Preview AMP' button

### PR DESCRIPTION
**Pull request by kienstra**
preview-amp-button

* This still needs more testing, but so far, it looks like it's working.
* It inserts the button into the DOM on `domReady`, and in that component's `componentDidUpdate()`, it moves it back to the correct position if needed.
* Re-rendering of the `<Header>` had moved the button, as it was inserted in the DOM, but this adds an operation to move it:

generates-amp-preview

This is a possible approach for PR #3323, so it's targeting that branch. For reference, here's full [diff against develop](https://github.com/ampproject/amp-wp/compare/develop...try/dom-manipulation-amp-preview).

Related to #2934

### CODE REVIEW

File: assets/css/src/amp-block-editor.css [hunk 1]
```diff
@@ -18,3 +18,35 @@
...
+}
+
+/*
+ * Remove right border radius on the (non-AMP) 'Preview' button, as it'll be next to the AMP button.
+ * @todo: only apply this if the 'Preview AMP' button displays.
+ */
+.wp-core-ui .editor-post-preview {
+	border-top-right-radius: 0;
+	border-bottom-right-radius: 0;
+}
```
Comment (kienstra) [inline assets/css/src/amp-block-editor.css]: @todo: this styling for the non-AMP preview button should only apply if there is 'Preview AMP' button:

non-amp-preview

File: assets/src/block-editor/plugins/amp-toggle.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 /**
  * External dependencies
  */
-import { errorMessages, isStandardMode } from 'amp-block-editor-data';
+import { errorMessages } from 'amp-block-editor-data';
```
Comment (kienstra) [inline assets/src/block-editor/plugins/amp-toggle.js]: @todo: there's a console error related to `amp-block-editor-data`, though probably not caused by this line:

```js
Uncaught ReferenceError: ampBlockEditor is not defined
    at Object.amp-block-editor-data (amp-block-validation.js:5180)
    at __webpack_require__ (amp-block-validation.js:20)
    at Module../assets/src/block-editor/components/amp-preview.js (amp-block-validation.js:114)
    at __webpack_require__ (amp-block-validation.js:20)
    at Module../assets/src/block-editor/components/index.js (amp-block-validation.js:471)
    at __webpack_require__ (amp-block-validation.js:20)
    at Module../assets/src/block-editor/helpers/index.js (amp-block-validation.js:907)
    at __webpack_require__ (amp-block-validation.js:20)
    at Module../assets/src/block-validation/index.js (amp-block-validation.js:2512)
    at __webpack_require__ (amp-block-validation.js:20)
```

File: assets/src/block-editor/components/amp-preview.js [hunk 3]
```diff
@@ -129,6 +139,23 @@ class AMPPreview extends Component {
...
 			this.setPreviewWindowLink( previewLink );
 		}
+
+		this.moveButton();
+	}
+
+	/**
+	 * Moves the (non-AMP) 'Preview' button to before this 'Preview AMP' button, if it's not there already.
+	 */
+	moveButton() {
```
Comment (kienstra) [inline assets/src/block-editor/components/amp-preview.js]: Calling this in the 'Preview AMP' button's `componentDidUpdate()` method moves it back to the correct place.

Without this, here's what it looks like:

amp-render-moved

File: assets/src/block-editor/helpers/index.js [hunk 4]
```diff
@@ -942,19 +942,22 @@ export const isAMPEnabled = () => {
...
 	}
 
 	const buttonWrapper = document.createElement( 'div' );
-	postPreviewButton.parentElement.insertBefore( buttonWrapper, postPreviewButton.nextSibling );
+	buttonWrapper.id = AMP_PREVIEW_BUTTON_WRAPPER_ID;
 
 	render(
 		<AMPPreview />,
 		buttonWrapper
 	);
```
Comment (kienstra) [inline assets/src/block-editor/helpers/index.js]: This renders the button into the DOM:

amp-preview-button

**Comment by kienstra** [comment]
**What do you think about this DOM rendering?**

Hi @swissspidy,
Hope you had a great weekend, and sorry for the delay.

Following up on https://github.com/ampproject/amp-wp/issues/2934#issuecomment-534629675, what do you think of this approach of rendering the 'Preview AMP' button into the DOM?

amp-preview-button

It's not ideal, as it could break with changes in Gutenberg. But as far as I've tested it, the button seems to work.

There are some `@todos` in this PR, it's not ready for a full review.

Thanks, Pascal!

**Comment by kienstra** [comment]
Hm...it looks like the button is unexpectedly moved on first creating a post and editing it.

**Comment by kienstra** [comment]
**Question About Overall Approach**

Hi @swissspidy,
Hope you're doing great.

What do you think of the overall approach here: [rendering the AMP preview button](https://github.com/ampproject/amp-wp/pull/3392/files#diff-3e719daf5951cbb5c3e722f91f8f40c2R956-R959) into the DOM with `render()`?


amp-preview-button

On `componentDidUpdate()`, it's [moved back into position](https://github.com/ampproject/amp-wp/pull/3392#discussion_r329799559).

This is a little brittle, as it could break with future versions of Gutenberg. But it works so far in my testing.

Thanks, Pascal!

**Comment by swissspidy** [comment]
@kienstra The approach is promising and the implementation looks reasonable. In terms of breakage, it doesn't seem to work on WordPress 5.3 RC (either with or without Gutenberg) with all these new styles.

**Comment by kienstra** [comment]
Thanks, Pascal! Good point, I'll look at how it works with WP 5.3 RC.

**Comment by kienstra** [comment]
Merging this `try` PR into the feature branch, but of course not into `develop` yet.

**Pull request merged**